master/pkg/searcher/asha.go

```diff
@@ -1,169 +1,181 @@
 type rung struct {
-	stepsNeeded   int
-	metrics       []trialMetric
-	startTrials   int
-	promoteTrials int
+	divisor           float64
+	stepsNeeded       int
+	metrics           []trialMetric
+	startTrials       int
+	outstandingTrials int
+	promoteTrials     int
 }
 
 // promotions handles bookkeeping of validation metrics and returns a RequestID to promote if
 // appropriate.
-func (r *rung) promotions(requestID RequestID, metric float64) []RequestID {
+func (r *rung) promotionsAsync(requestID RequestID, metric float64) []RequestID {
+	// See if there is a trial to promote. We are increasing the total number of trials seen by 1; the
+	// number of best trials that definitely should have been promoted so far (numPromote) can only
+	// stay the same or increase by 1.
+	oldNumPromote := int(float64(len(r.metrics)) / r.divisor)
+	numPromote := int(float64(len(r.metrics)+1) / r.divisor)
+
 	// Insert the new trial result in the appropriate place in the sorted list.
 	insertIndex := sort.Search(
 		len(r.metrics),
 		func(i int) bool { return r.metrics[i].metric > metric },
 	)
+	promoteNow := insertIndex < numPromote
+
 	r.metrics = append(r.metrics, trialMetric{})
 	copy(r.metrics[insertIndex+1:], r.metrics[insertIndex:])
 	r.metrics[insertIndex] = trialMetric{
 		requestID: requestID,
 		metric:    metric,
+		promoted:  promoteNow,
+		closed:    false,
 	}
 
-	// If there are enough trials done to definitively promote one, do so. Otherwise, return nil.
-	currPromote := len(r.metrics) + r.promoteTrials - r.startTrials
-	switch {
-	case currPromote <= 0: // Not enough trials completed for any promotions.
-		return nil
-	case insertIndex < currPromote: // Incoming trial should be promoted.
+	// If the new trial is good enough, it should be promoted immediately (whether or not numPromote
+	// changes). Otherwise, if numPromote changes, there is some other trial that should be promoted,
+	// unless it has been promoted already.
+	if promoteNow {
 		return []RequestID{requestID}
-	default: // Promote next trial in sorted metrics array.
-		t := &r.metrics[currPromote-1]
-		return []RequestID{t.requestID}
 	}
+	if numPromote != oldNumPromote {
+		t := &r.metrics[oldNumPromote]
+		if !t.promoted {
+			t.promoted = true
+			return []RequestID{t.requestID}
+		}
+	}
+
+	return nil
 }
 
 func (s *asyncHalvingSearch) initialOperations(ctx context) ([]Operation, error) {
+	// We can initialize bottom rung with divisor^(numRung-1) configurations since
+	// it will take that many configurations on average to be able to promote one
+	// configuration to the top rung.
 	var ops []Operation
 	for trial := 0; trial < s.rungs[0].startTrials; trial++ {
 		create := NewCreate(
 			ctx.rand, sampleAll(ctx.hparams, ctx.rand), model.TrialWorkloadSequencerType)
 		ops = append(ops, create)
 		ops = append(ops, trainAndValidate(create.RequestID, 0, s.rungs[0].stepsNeeded)...)
+		s.rungs[0].outstandingTrials++
+		s.trialRungs[create.RequestID] = 0
```

Comment:
question: why is this now needed?

Reply:
trialRungs for a request needs to be initialized at the same time as trial creation since it is needed as part of validationCompleted.  oustandingTrials is used to track pending jobs so we know when all the jobs in a rung are complete and can close them and ultimately finish the experiment.